src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingSourceGenerator.cs

```diff
@@ -0,0 +1,194 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+//#define LAUNCH_DEBUGGER
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.DotnetRuntime.Extensions;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
+{
+    /// <summary>
+    /// Generates source code to optimize binding with ConfigurationBinder.
+    /// </summary>
+    [Generator]
+    public sealed partial class ConfigurationBindingSourceGenerator : IIncrementalGenerator
+    {
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            IncrementalValueProvider<KnownTypeData> compilationData =
+                context.CompilationProvider
+                    .Select((compilation, _) => new KnownTypeData(compilation));
+
+            IncrementalValuesProvider<BinderInvocationOperation> inputCalls = context.SyntaxProvider.CreateSyntaxProvider(
+                (node, _) => node is InvocationExpressionSyntax invocation,
+                (context, cancellationToken) => new BinderInvocationOperation(context, cancellationToken));
+
+            IncrementalValueProvider<(KnownTypeData, ImmutableArray<BinderInvocationOperation>)> inputData = compilationData.Combine(inputCalls.Collect());
```

Comment:
Overall shape of the the incremental pipeline looks good. In order to be able to utilize caching you still need to provide an equality comparer for the intermediate `(KnownTypeData, ImmutableArray<BinderInvocationOperation>)` type. You can pass an equality comparer using the following change:
```suggestion
            IncrementalValueProvider<(KnownTypeData, ImmutableArray<BinderInvocationOperation>)> inputData = compilationData.Combine(inputCalls.Collect()).WithComparer(new MyIntermediateValuesComparer());
```

See also https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md#comparing-items for more details.

Reply:
Thanks. Per our private conversation I'll address this with high priority in a follow up since it is a bit involved but doesn't block getting the feature out to users.